src/LinearOperators.jl

```diff
@@ -85,346 +90,355 @@ end
 symmetric."""
-LinearOperator(M :: SymTridiagonal) =
-    LinearOperator(M; symmetric = true, hermitian = eltype(M) <: Real)
+LinearOperator{T}(M :: SymTridiagonal{T}) =
+    LinearOperator(M; symmetric=true, hermitian=eltype(M) <: Real)
 
+# the only advantage of this constructor is optional args
+# use LinearOperator{Float64} if you mean real instead of complex
 "Construct a linear operator from functions."
-LinearOperator(nrow :: Int, ncol :: Int, dtype :: DataType,
-               symmetric :: Bool, hermitian :: Bool,
-               prod :: Function, tprod :: Function, ctprod :: Function) =
-  LinearOperator(nrow, ncol, dtype, symmetric, hermitian,
-                 prod, Nullable{Function}(tprod), Nullable{Function}(ctprod))
-
-"Construct a real symmetric linear operator from a function."
-LinearOperator(nrow :: Int, dtype :: DataType, prod :: Function) =
-  LinearOperator(nrow, nrow, dtype, true, true,
-                 prod,
-                 Nullable{Function}(prod),
-                 Nullable{Function}(prod))
-
-"Construct a linear operator from a single function."
-LinearOperator(nrow :: Int, ncol :: Int, dtype :: DataType,
-               symmetric :: Bool, hermitian :: Bool,
-               prod :: Function) =
-  LinearOperator(nrow, ncol, dtype, symmetric, hermitian,
-                 prod, Nullable{Function}(), Nullable{Function}())
-
-
-if VERSION ≥ v"0.4.0-dev"
-  import Base.+, Base.-, Base.*, Base.(.+), Base.(.-), Base.(.*)
-  import Base.transpose, Base.ctranspose
+function LinearOperator(nrow :: Int, ncol :: Int,
+                        symmetric :: Bool, hermitian :: Bool,
+                        prod :: Function,
+                        tprod :: Union{Function,Nullable{Function}}=Nullable{Function}(),
+                        ctprod :: Union{Function,Nullable{Function}}=Nullable{Function}())
+
+  T = hermitian ? (symmetric ? Float64 : Complex128) : Complex128
+  LinearOperator{T}(nrow, ncol, symmetric, hermitian, prod, tprod, ctprod)
 end
 
+# "Construct a real symmetric or complex hermitian linear operator from a function."
+# LinearOperator(nrow :: Int, T :: DataType, prod :: Function) =
+#   LinearOperator{T}(nrow, nrow, T <: Real, T <: Real || T <: Complex,
+#                     prod, Nullable{Function}(prod), Nullable{Function}(prod))
+
+# "Construct a linear operator from a single function."
+# LinearOperator(nrow :: Int, ncol :: Int,
+#                symmetric :: Bool, hermitian :: Bool,
+#                prod :: Function) =
+#   LinearOperator(nrow, ncol, symmetric, hermitian,
+#                  prod, Nullable{Function}(), Nullable{Function}())
+
+
 # Apply an operator to a vector.
-function (*)(op :: AbstractLinearOperator, v :: Vector)
-  (m, n) = size(op)
-  if size(v,1) != n
-    error("Shape mismatch")
-  end
-  return op.prod(v)
+function *(op :: AbstractLinearOperator, v :: Vector)
+  size(v, 1) == size(op, 2) || throw(LinearOperatorException("shape mismatch"))
+  op.prod(v)
 end
```

Comment:
Without investigating much further, it appears to be so. The checks and promotes are mostly to create new operators.
LBFGS sees more use, because the data is created with that type, but even so, we may be better off without it.
It may be useful for different Floats (or BigFloats), possibly, but it can't be used with Integers, because of the square roots, or with Complex because of the `dot(s,y) <= 1.0e-20`.